nylon-threads/src/main/java/com/palantir/nylon/threads/VirtualThreads.java

```diff
@@ -0,0 +1,276 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.nylon.threads;
+
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.lang.Thread.UncaughtExceptionHandler;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Utility functionality to allow libraries which compile for earlier JDKs
+ * to take advantage of virtual threads features on sufficiently new runtimes.
+ */
+public final class VirtualThreads {
+    private static final SafeLogger log = SafeLoggerFactory.get(VirtualThreads.class);
+
+    @Nullable
+    private static final VirtualThreadSupport VIRTUAL_THREAD_SUPPORT = maybeInitialize();
+
+    public static boolean isVirtual(Thread thread) {
+        return VIRTUAL_THREAD_SUPPORT != null && VIRTUAL_THREAD_SUPPORT.isVirtual(thread);
+    }
+
+    public static Optional<VirtualThreadSupport> get() {
+        return Optional.ofNullable(VIRTUAL_THREAD_SUPPORT);
+    }
+
+    @Nullable
+    private static VirtualThreadSupport maybeInitialize() {
+        int featureVersion = Runtime.version().feature();
+        if (featureVersion < 21) {
+            if (log.isDebugEnabled()) {
+                log.debug(
+                        "Virtual threads are not available prior to jdk21",
+                        SafeArg.of("currentVersion", featureVersion));
+            }
+            return null;
+        }
+        try {
+            return new ReflectiveVirtualThreadSupport();
+        } catch (Throwable t) {
+            log.warn("Virtual thread support is not available", t);
+            return null;
+        }
+    }
+
+    private VirtualThreads() {}
+
+    public interface VirtualThreadSupport {
+        /**
+         * Returns {@code true} if {@code thread} is virtual, {@code false} otherwise.
+         * Equivalent to {@code thread.isVirtual()} on jdk-21+.
+         */
+        boolean isVirtual(Thread thread);
+
+        /** Equivalent to {@code Executors.newThreadPerTaskExecutor(threadFactory)} on jdk-21+. */
+        ExecutorService newThreadPerTaskExecutor(ThreadFactory threadFactory);
+
+        /** Alias to the jdk-21 {@code Thread.ofVirtual()} API. */
+        VirtualThreadBuilder ofVirtual();
+
+        interface VirtualThreadBuilder {
```

Comment:
should we add JavaDoc or link https://download.java.net/java/early_access/jdk21/docs/api/java.base/java/lang/Thread.Builder.html?

Reply:
Good call. I've pointed the documentation at the `OfVirtual` subtype found here: https://download.java.net/java/early_access/jdk21/docs/api/java.base/java/lang/Thread.Builder.OfVirtual.html

I've used the GA javadoc url format and pointed to jdk21, which has not been published yet, but I suspect it will be more durable once it goes live. Let me know if you'd prefer the early_access page that's guaranteed to use the correct format for now.